crates/types/src/nodes_config.rs

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+// Mute clippy until this file is used.
+#![allow(dead_code)]
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+use enumset::{EnumSet, EnumSetType};
+
+use crate::{GenerationalNodeId, NodeId, PlainNodeId};
+
+#[derive(Debug, thiserror::Error)]
+pub enum NodesConfigError {
+    #[error("node {0} was not found in config")]
+    UnknownNodeId(NodeId),
+    #[error("node {0} has been permanently deleted from config")]
+    Deleted(NodeId),
+    #[error("node was found but has a mismatching generation (expected={expected} found={found})")]
+    GenerationMismatch { expected: NodeId, found: NodeId },
+}
+
+// PartialEq+Eq+Clone+Copy are implemented by EnumSetType
+#[derive(Debug, Hash, EnumSetType, derive_more::Display)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Role {
+    Worker,
+    ClusterContoller,
+}
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    Default,
+    derive_more::Display,
+    derive_more::From,
+    derive_more::Into,
+    derive_more::AddAssign,
+)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[display(fmt = "v{}", _0)]
+pub struct ConfigVersion(u32);
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+struct NodesConfiguration {
+    version: ConfigVersion,
+    nodes: HashMap<PlainNodeId, MaybeNode>,
+    controllers: Vec<ControllerConfig>,
+    name_lookup: HashMap<String, PlainNodeId>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, strum_macros::EnumIs)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+enum MaybeNode {
+    Tombstone,
+    Node(NodeConfig),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+struct NodeConfig {
+    pub name: String,
```

Comment:
Are names a convenience feature for users to specify a name for a node?

Reply:
node names are what the user actually provides on startup and attach should return a generational node id. Environments like k8s will give us a stable name which controller can use to locate the node and bump the generation on attach. I'll make the change in a follow up commit to do that.

Reply:
Alright. So far I was assuming that the `NodeId` would be the stable identifier for a node.